workshop1/slides/src/slides.md

```diff
@@ -0,0 +1,664 @@
+  * Takes two inputs: a function and a list.
+  * Applies the function to every item in the list
+
+```haskell
+map :: (a -> b) -> [a] -> [b]
+map f []     = []
+map f (x:xs) = f x : map f xs
+```
+
+* Defined using *recursion*
+
+
+## Writing `parseFile :: CSV -> DataModel`
+
+* We are ready to write the full parser function:
+
+```haskell
+type CSV = String
+type DataModel = [Maybe Candidate]
+
+parseFile :: CSV -> DataModel
+parseFile = undefined
+```
+
+* `lines` splits strings into lines
+* `tail` removes the first item in a list
+* Try and use point-free style
+
+## Writing `parseFile :: CSV -> DataModel`
+
+Our solution:
+
+```haskell
+parseFile :: CSV -> DataModel
+parseFile = map parseLine . tail . lines
+```
+
+
+## Generating `encode :: DataModel -> JSON`
+
+```haskell
+{-# LANGUAGE DeriveGeneric #-}
+main Main where
+
+import           GHC.Generics
+import qualified Data.Aeson      as JSON
+import qualified Data.ByteString as B
+
+data Candidate = Candidate
+  { ...
+  } deriving (Show, Generic)
+
+instance Aeson.ToJSON Candidate
+```
+
+* Language extensions expand the Haskell language
+* `Data.Aeson` provides encoding to and from JSON
+* `Data.ByteString` provides efficient binary strings
+
+## Generating `encode :: DataModel -> JSON`
+
+```haskell
+type JSON = ByteString
+```
+
+* `encode` outputs efficient binary strings
+* Try it out in the REPL!
+
+![](images/encoding.png)
+
+# Connecting to the Outside
+
+## Goal
+
+* Write a complete program!
+* Introduce IO and do notation
+
+### Approach
+
+* Introduce purity and impurity
+* Write main function
+
+![](images/connecting.pdf)
+
+## Purity in Haskell
+
+* All functions we have written in the previous section are *pure*
+* A pure function **always** returns the same *output* when given the same *input*
+* This behaviour is *enforced by the compiler*
+* Why?
+  * Code is easier to refactor and reason about
+  * Interactions with the outside world are more explicit
+  * Easy to test
+* Pure functions cannot interact with the world
+* So how do we communicate with the outside world?
+
+
+## IO
+
+* Communication with the outside handled through IO
+* Impurity is **always** explicitly encoded in the types
+
+![](images/io.pdf){ width=100% }
+
+
+## Do notation
+
+```haskell
+askName :: IO ()
+askName = do
+  putStrLn "Enter your name:"
+  name <- getLine
+  let reversedName = reverse name
+  putStrLn ("Your name reversed: " ++ reversedName)
+```
+
+* `putStrLn` prints a string
+* `getLine` gets a line of input from the user
+* Do blocks are a special Haskell syntax impure programming
+  * Always start with the `do` keyword
+  * Functions executed top-to-bottom
+  * Return values stored in local variables using `(<-)`
+  * `let` used for pure computation
+
+
+## Read / Write
+
+```haskell
+convert :: String -> IO ()
+convert filename = do
+  undefined
+```
+
+* `readFile` returns the contents of a file as a String
+* `B.writeFile` writes a ByteString to a file
```

Comment:
We need give emphasis on the qualified import of the *lazy* bytestring above

Reply:
:+1:

Reply:
Figured we can do this out loud